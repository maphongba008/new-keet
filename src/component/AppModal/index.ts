import AppModal from './AppModal'
import * as ModalTypes from './types'

export default AppModal
export { ModalTypes }
