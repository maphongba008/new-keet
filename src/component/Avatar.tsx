import React, { memo, ReactNode, useMemo } from 'react'
import {
  ImageStyle,
  StyleProp,
  StyleSheet,
  Text,
  View,
  ViewProps,
  ViewStyle,
} from 'react-native'
import { Image } from 'expo-image'

import { getFirstLetters } from '@holepunchto/keet-store/store/lib/string'

import { APPIUM_IDs, appiumTestProps } from 'lib/appium'
import { getScaledFontSize, randomColor } from 'lib/avatar'
import s, {
  ICON_SIZE_72,
  UI_SIZE_2,
  UI_SIZE_12,
  UI_SIZE_32,
} from 'lib/commonStyles'
import { MemberType } from 'lib/types'

import SvgIcon from './SvgIcon'
import { colors, createThemedStylesheet } from './theme'

// https://github.com/holepunchto/settings/blob/651dcd959994fc1f270f2da5ffbf37fb3fa0ba2e/settings.js#L150

export interface AvatarProps {
  base64?: string | null
  seed?: string
  small?: boolean
  style?: ViewProps['style']
  name?: string | null
  color?: string | null
}

export interface AvatarContainerProps {
  children?: ReactNode
  small?: boolean
  style?: ViewProps['style']
}

export const AvatarContainer: React.FC<AvatarContainerProps> = ({
  small,
  style,
  children,
}) => {
  const styles = getStyles()
  const containerStyle = useMemo(() => {
    if (small) {
      return [style, styles.smallContainer]
    }

    return [style]
  }, [small, style, styles.smallContainer])

  return <View style={containerStyle}>{children}</View>
}

export const Avatar: React.FC<AvatarProps> = memo(
  ({ base64: uri, seed, small, style, color, name }) => {
    const styles = getStyles()

    const avatarDisplayName = getFirstLetters(name ?? '').substring(0, 2)
    const avatarStyle = useMemo(() => {
      const avatarStyles: StyleProp<ViewStyle> = []

      if (!uri) {
        avatarStyles.push(s.centeredLayout, {
          backgroundColor: color || randomColor(seed),
        })
      }

      if (small) {
        avatarStyles.push(styles.smallImg, style)
      } else {
        avatarStyles.push(styles.bigImg, style)
      }

      return StyleSheet.flatten(avatarStyles)
    }, [color, seed, small, style, styles.bigImg, styles.smallImg, uri])

    const circleSize = small ? 16 : 40

    if (uri) {
      return (
        <Image
          style={avatarStyle as ImageStyle}
          source={{ uri }}
          contentFit="contain"
          recyclingKey={uri}
          cachePolicy="memory"
          {...appiumTestProps(APPIUM_IDs.avatar_image)}
        />
      )
    }

    if (avatarDisplayName) {
      return (
        <View style={avatarStyle} {...appiumTestProps(APPIUM_IDs.avatar_image)}>
          <Text
            style={[
              styles.avatarText,
              {
                fontSize: getScaledFontSize(avatarStyle),
              },
            ]}
          >
            {avatarDisplayName}
          </Text>
        </View>
      )
    }
    return (
      <View style={avatarStyle} {...appiumTestProps(APPIUM_IDs.avatar_image)}>
        <SvgIcon
          name="keetOutline"
          width={circleSize}
          height={circleSize}
          color={colors.white_snow}
        />
      </View>
    )
  },
)

type MemberAvatarProps = {
  member: Partial<MemberType>
} & AvatarProps

export const MemberAvatar = ({ member, ...props }: MemberAvatarProps) => {
  const base64 = useMemo(() => {
    return member?.blocked ? null : member?.avatarUrl
  }, [member])

  return (
    <Avatar
      name={member.displayName}
      small
      base64={base64}
      color={member?.color}
      seed={member?.swarmId}
      {...props}
    />
  )
}

export const AVATAR_SIZE = UI_SIZE_32 - UI_SIZE_2
export const AVATAR_SIZE_BIG = ICON_SIZE_72
export const LOBBY_AVATAR_SIZE = 55

interface AvatarStyleSizeParams {
  size: number
  borderWidth: number
  padding: number
}

const getStyles = createThemedStylesheet((theme) => {
  const smallPreset: AvatarStyleSizeParams = {
    size: AVATAR_SIZE,
    borderWidth: 1,
    padding: 1,
  }
  const bigPreset: AvatarStyleSizeParams = {
    size: AVATAR_SIZE_BIG,
    borderWidth: 2,
    padding: 1,
  }

  const getContainerStyle = ({
    size,
    borderWidth,
    padding,
  }: AvatarStyleSizeParams) => ({
    borderRadius: size / 2,
    borderWidth,
    padding,
    borderColor: theme.color.red_400,
    width: size,
    height: size,
  })

  const getImgStyle = (
    { size, borderWidth, padding }: AvatarStyleSizeParams,
    isHighlighted?: boolean,
  ) => {
    const sizeInset = isHighlighted ? size - 2 * (borderWidth + padding) : size

    return {
      width: sizeInset,
      height: sizeInset,
      borderRadius: sizeInset / 2,
    }
  }

  const styles = StyleSheet.create({
    avatarText: {
      ...theme.text.bodyBold,
      fontSize: UI_SIZE_12,
    },
    bigImg: getImgStyle(bigPreset),
    smallContainer: getContainerStyle(smallPreset),
    smallImg: getImgStyle(smallPreset),
  })
  return styles
})
