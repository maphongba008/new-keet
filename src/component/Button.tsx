import React, { memo, useCallback, useEffect } from 'react'
import {
  Pressable,
  PressableProps,
  StyleProp,
  StyleSheet,
  Text,
  View,
  ViewStyle,
  type ColorValue,
  type ViewProps,
} from 'react-native'
import isEqual from 'react-fast-compare'
import Reanimated, {
  useAnimatedStyle,
  useSharedValue,
  withTiming,
  type AnimatedProps,
} from 'react-native-reanimated'

import { appiumTestProps } from 'lib/appium'
import s, {
  DIRECTION_CODE,
  ICON_SIZE_16,
  ICON_SIZE_20,
  TRANSPARENT,
  UI_SIZE_8,
} from 'lib/commonStyles'

import {
  colorWithAlpha,
  createThemedStylesheet,
  hexToRgbOpacity,
  useTheme,
} from './theme'

export type ButtonBaseProps = Omit<AnimatedProps<ViewProps>, 'testID'> & {
  hint?: string
  disableFade?: boolean
  enabled?: boolean
  onPress?: () => void
  delayLongPress?: number
  onLongPress?: () => void
  disabled?: boolean
  testID?: string
}

export const AnimatedPressable = Reanimated.createAnimatedComponent(Pressable)

export const ButtonBase = memo((props: ButtonBaseProps) => {
  const { hint, disableFade, style, disabled, ...rest } = props
  const theme = useTheme()

  const opacity = useSharedValue(disabled ? 0.5 : 1)
  const duration = theme.animation.ms

  useEffect(() => {
    opacity.value = withTiming(disabled ? 0.5 : 1, { duration })
  }, [disabled, opacity, duration])

  const animatedStyles = useAnimatedStyle(() => ({ opacity: opacity.value }))

  const onPressIn = useCallback(() => {
    if (disableFade === true) {
      return
    }
    opacity.value = withTiming(0.64, { duration })
  }, [disableFade, duration, opacity])

  const onPressOut = useCallback(() => {
    if (disableFade === true) {
      return
    }
    opacity.value = withTiming(1, { duration })
  }, [disableFade, duration, opacity])

  const animatedPressableProps = {
    accessible: true,
    accessibilityHint: hint,
    accessibilityRole: 'button',
    unstable_pressDelay: 5,
    ...appiumTestProps(props.testID),
    onPressIn,
    onPressOut,
    disabled,
    style: [animatedStyles, style],
    ...rest,
  } as AnimatedProps<PressableProps>

  return <AnimatedPressable {...animatedPressableProps} />
}, isEqual)

export const IconButton = memo((props: ButtonBaseProps) => {
  const { style, ...rest } = props
  const styles = getStyles()

  return (
    <ButtonBase
      {...rest}
      style={[s.centeredLayout, styles.buttonBase, style]}
    />
  )
}, isEqual)

export enum TextButtonType {
  link,
  cancel,
  primary,
  primaryOutline,
  primaryTransparentBg,
  secondary,
  gray,
  disabled,
  danger,
  outline,
  dangerTransparentBg,
  secondaryDanger,
}
export const TextButton = (
  props: ButtonBaseProps & {
    text: string
    icon?: React.FC
    type?: TextButtonType
  },
) => {
  const { style, text, type, ...rest } = props
  const styles = getStyles()

  let buttonStyle = {}
  let buttonTextStyle = {}
  switch (type) {
    case TextButtonType.link:
      buttonStyle = styles.textButton__link
      buttonTextStyle = styles.textButtonText__link
      break
    case TextButtonType.cancel:
      buttonStyle = { ...styles.textButton, ...styles.textButton__cancel }
      buttonTextStyle = {
        ...styles.textButtonText,
        ...styles.textButtonText__cancel,
      }
      break
    case TextButtonType.gray:
      buttonStyle = { ...styles.textButton, ...styles.textButton__gray }
      buttonTextStyle = {
        ...styles.textButtonText,
        ...styles.textButtonText__gray,
      }
      break
    case TextButtonType.primary:
      buttonStyle = { ...styles.textButton, ...styles.textButton__primary }
      buttonTextStyle = {
        ...styles.textButtonText,
        ...styles.textButtonText__primary,
      }
      break
    case TextButtonType.primaryOutline:
      buttonStyle = {
        ...styles.textButton,
        ...styles.textButton__primary_outline,
      }
      buttonTextStyle = {
        ...styles.textButtonText,
        ...styles.textButtonText__primary_outline,
      }
      break
    case TextButtonType.secondary:
      buttonStyle = {
        ...styles.textButton,
        ...styles.textButtonNoBorder,
      }
      buttonTextStyle = {
        ...styles.textButtonText,
        ...styles.textButtonNoTransform,
      }
      break
    case TextButtonType.secondaryDanger:
      buttonStyle = {
        ...styles.textButton,
        ...styles.textButton__secondary_danger,
      }
      buttonTextStyle = {
        ...styles.textButtonText,
        ...styles.textButtonText__secondary_danger,
      }
      break
    case TextButtonType.primaryTransparentBg:
      buttonStyle = {
        ...styles.textButton,
        ...styles.textButton__primary_outline,
        ...styles.textButtonTransparentBg,
      }
      buttonTextStyle = {
        ...styles.textButtonText,
        ...styles.textButtonText__primary_outline,
      }
      break
    case TextButtonType.dangerTransparentBg:
      buttonStyle = {
        ...styles.textButton,
        ...styles.textButton__danger_outline,
        ...styles.textButtonTransparentBg,
      }
      buttonTextStyle = {
        ...styles.textButtonText,
        ...styles.textButtonText__danger_outline,
      }
      break
    case TextButtonType.disabled:
      buttonStyle = { ...styles.textButton, ...styles.textButton__disabled }
      buttonTextStyle = {
        ...styles.textButtonText,
        ...styles.textButtonText__disabled,
      }
      break
    case TextButtonType.danger:
      buttonStyle = { ...styles.textButton, ...styles.textButton__danger }
      buttonTextStyle = {
        ...styles.textButtonText,
        ...styles.textButtonText__danger,
      }
      break
    case TextButtonType.outline:
      buttonStyle = {
        ...styles.textButton,
        ...styles.textButton__outline,
      }
      buttonTextStyle = {
        ...styles.textButtonText,
        ...styles.textButtonText__outline,
      }
      break
    default:
      buttonStyle = styles.textButton
      buttonTextStyle = styles.textButtonText
      break
  }

  return (
    <ButtonBase {...rest} style={[buttonStyle, style]}>
      {props.icon !== undefined && <props.icon />}
      <Text style={buttonTextStyle}>{text}</Text>
    </ButtonBase>
  )
}

export const OptionButton = (
  props: ButtonBaseProps & { text: string; selected: boolean; textStyle?: any },
) => {
  const { style, textStyle, text, selected, ...rest } = props
  const styles = getStyles()

  return (
    <ButtonBase
      {...rest}
      style={[
        style,
        selected ? styles.optionButtonSelected : styles.optionButton,
      ]}
    >
      <Text
        style={[
          selected ? styles.optionButtonSelectedText : styles.optionButtonText,
          textStyle,
        ]}
      >
        {text}
      </Text>
    </ButtonBase>
  )
}

export interface OutlineButtonIconProps {
  size?: number
  color?: ColorValue
  style: StyleProp<ViewStyle>
}

export interface OutlineButtonProps {
  title: string
  color?: string
  iconColor?: string
  subtitle?: string
  PrimaryIcon?: React.FC<OutlineButtonIconProps>
  SecondaryIcon?: React.FC<OutlineButtonIconProps>
}

export const OutlineButton = (props: ButtonBaseProps & OutlineButtonProps) => {
  const {
    style,
    iconColor,
    color,
    title,
    subtitle,
    PrimaryIcon,
    SecondaryIcon,
    ...rest
  } = props
  const styles = getStyles()

  return (
    <ButtonBase {...rest} style={[styles.outlineButton, style]}>
      <View style={[s.container, s.row, s.alignItemsCenter, s.justifyStart]}>
        {!(PrimaryIcon == null) && (
          <PrimaryIcon
            size={styles.iconStart.width}
            color={iconColor ?? color ?? styles.outlineButtonText.color}
            style={styles.iconStart}
          />
        )}
        <Text
          style={[
            styles.title,
            s.container,
            // eslint-disable-next-line react-native/no-inline-styles
            {
              color: color ?? styles.outlineButtonText.color,
              lineHeight: PrimaryIcon != null ? 21 : 22,
            },
          ]}
          numberOfLines={1}
          ellipsizeMode="tail"
        >
          {title}
        </Text>
      </View>
      <View style={[s.container, s.row, s.alignItemsCenter, s.justifyEnd]}>
        <Text style={styles.subtitle} numberOfLines={1}>
          {subtitle}
        </Text>
        {!(SecondaryIcon == null) && (
          <SecondaryIcon
            size={styles.iconEnd.width}
            color={color ?? (styles.outlineButtonText.color as string)}
            style={styles.iconEnd}
          />
        )}
      </View>
    </ButtonBase>
  )
}

const getStyles = createThemedStylesheet((theme) => {
  const styles = StyleSheet.create({
    buttonBase: {
      height: theme.icon.size,
      width: theme.icon.size,
    },
    iconEnd: {
      height: ICON_SIZE_16,
      marginStart: theme.spacing.standard,
      width: ICON_SIZE_16,
    },
    iconStart: {
      height: ICON_SIZE_20,
      marginEnd: theme.spacing.standard,
      width: ICON_SIZE_20,
    },
    optionButton: {
      ...s.centeredLayout,
      backgroundColor: theme.color.bg4,
      borderRadius: theme.border.radiusLarge,
      padding: UI_SIZE_8,
    },
    optionButtonSelected: {
      ...s.centeredLayout,
      backgroundColor: theme.color.blue_400,
      borderRadius: theme.border.radiusLarge,
      padding: UI_SIZE_8,
    },
    optionButtonSelectedText: {
      ...theme.text.body,
      color: theme.color.grey_900,
      ...s.textAlignCenter,
    },
    optionButtonText: {
      ...theme.text.body,
      ...s.textAlignCenter,
    },
    outlineButton: {
      ...s.row,
      ...s.alignItemsCenter,
      ...s.flexSpaceBetween,
      backgroundColor: theme.background.bg_1,
      borderColor: theme.border.color,
      borderRadius: 12,
      borderWidth: theme.border.width,
      height: 54,
      padding: theme.spacing.standard,
    },
    // eslint-disable-next-line react-native/no-unused-styles
    outlineButtonText: {
      color: theme.text.body.color,
    },
    subtitle: {
      ...theme.text.body,
      opacity: 0.4,
      writingDirection: DIRECTION_CODE,
    },
    textButton: {
      backgroundColor: colorWithAlpha(theme.color.blue_400, 0.2),
      borderRadius: theme.border.radiusNormal,
      paddingVertical: theme.spacing.standard,
      ...s.row,
      ...s.centeredLayout,
    },
    textButtonNoBorder: {
      borderWidth: 0,
    },
    textButtonNoTransform: {
      textTransform: 'none',
    },
    textButtonText: {
      ...theme.text.bodySemiBold,
      color: theme.color.blue_400,
      marginHorizontal: theme.spacing.standard / 2,
      ...s.uppercase,
      ...s.textAlignCenter,
    },
    textButtonText__cancel: {
      color: theme.text.bodyBold.color,
      textTransform: 'none',
    },
    textButtonText__danger: {
      ...theme.text.bodySemiBold,
      color: theme.color.red_400,
      textTransform: 'none',
    },
    textButtonText__danger_outline: {
      ...theme.text.bodySemiBold,
      color: theme.color.danger,
      textTransform: 'none',
    },
    textButtonText__disabled: {
      ...theme.text.bodySemiBold,
      color: theme.color.grey_200,
      textTransform: 'none',
    },
    textButtonText__gray: {
      ...theme.text.bodySemiBold,
      textTransform: 'none',
    },
    textButtonText__link: {
      ...theme.text.bodyBold,
      ...s.uppercase,
      ...s.textAlignCenter,
      color: theme.color.accent,
      marginHorizontal: theme.spacing.standard / 2,
    },
    textButtonText__outline: {
      ...theme.text.bodySemiBold,
      color: theme.color.blue_400,
      textTransform: 'none',
    },
    textButtonText__primary: {
      ...theme.text.bodySemiBold,
      color: theme.color.bg,
      textTransform: 'none',
    },
    textButtonText__primary_outline: {
      ...theme.text.bodySemiBold,
      color: theme.color.blue_400,
      textTransform: 'none',
    },
    textButtonText__secondary_danger: {
      ...theme.text.bodySemiBold,
      color: theme.color.red_400,
      textTransform: 'none',
    },
    textButtonTransparentBg: {
      backgroundColor: TRANSPARENT,
    },
    textButton__cancel: {
      backgroundColor: theme.color.grey_600,
      borderColor: theme.color.grey_400,
      borderWidth: theme.border.width,
    },
    textButton__danger: {
      backgroundColor: colorWithAlpha(theme.color.red_400, 0.2),
    },
    textButton__danger_outline: {
      ...theme.text.bodySemiBold,
      borderColor: theme.color.danger,
      borderWidth: theme.border.width,
      color: theme.color.danger,
      textTransform: 'none',
    },
    textButton__disabled: {
      backgroundColor: theme.color.grey_400,
    },
    textButton__gray: {
      backgroundColor: theme.color.grey_600,
    },
    textButton__link: {
      ...s.row,
      ...s.centeredLayout,
      paddingVertical: theme.spacing.standard / 2,
    },
    textButton__outline: {
      backgroundColor: TRANSPARENT,
      borderColor: colorWithAlpha(theme.color.blue_400, 0.2),
      borderWidth: theme.border.width,
    },
    textButton__primary: {
      backgroundColor: theme.color.blue_400,
    },
    textButton__primary_outline: {
      backgroundColor: hexToRgbOpacity(theme.color.blue_400, 0.2),
      borderColor: theme.color.blue_400,
      borderWidth: theme.border.width,
    },
    textButton__secondary_danger: {
      backgroundColor: theme.color.red_900,
    },
    title: {
      ...theme.text.body,
      margin: -2,
      minWidth: '70%',
      writingDirection: DIRECTION_CODE,
    },
  })
  return styles
})
