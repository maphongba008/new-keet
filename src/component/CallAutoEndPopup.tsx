import React, { memo, useCallback, useEffect, useMemo } from 'react'
import { StyleSheet, Text, View } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'
import { Notifier } from 'react-native-notifier'
import { SafeAreaView } from 'react-native-safe-area-context'

import {
  getAppCurrentCallRoomId,
  getAppCurrentRoomId,
} from '@holepunchto/keet-store/store/app'
import {
  getCallAutoEndCountdown,
  getCallShowAutoEndPopup,
  getCallStatus,
  isCallJoinedMode,
  resetCallAutoEndTimer,
} from '@holepunchto/keet-store/store/call'
import { switchRoomSubmit } from '@holepunchto/keet-store/store/room'

import {
  closeBottomSheet,
  showBottomSheet,
} from 'component/AppBottomSheet/AppBottomSheet.Store'
import BottomSheetEnum from 'component/AppBottomSheet/SheetComponents/BottomSheetEnum'
import s, {
  ICON_SIZE_16,
  UI_SIZE_4,
  UI_SIZE_8,
  UI_SIZE_12,
  UI_SIZE_16,
} from 'lib/commonStyles'
import { useConfig } from 'lib/hooks/useRoom'
import { showInfoNotifier } from 'lib/hud'
import { getCurrentRoute, navigate, SCREEN_ROOM_CALL } from 'lib/navigation'

import { useStrings } from 'i18n/strings'

import { CloseButton } from './CloseButton'
import { RoomAvatarImage } from './RoomAvatarImage'
import SvgIcon from './SvgIcon'
import { colors, createThemedStylesheet } from './theme'

function useCallPanelOpened() {
  const callStatus = useSelector(getCallStatus)
  const isCallJoined = isCallJoinedMode(callStatus)
  const callRoomId = useSelector(getAppCurrentCallRoomId)
  const roomId = useSelector(getAppCurrentRoomId)
  const route = getCurrentRoute()
  return isCallJoined && callRoomId === roomId && route === 'call'
}

function useCountdown() {
  const autoEndCountdown = useSelector(getCallAutoEndCountdown)
  return useMemo(
    () =>
      Math.round((isFinite(autoEndCountdown) ? autoEndCountdown : 0) / 60000),
    [autoEndCountdown],
  )
}

const CallAutoEndPopup = memo(() => {
  const dispatch = useDispatch()
  const callOpened = useCallPanelOpened()
  const showAutoEndPopup = useSelector(getCallShowAutoEndPopup)
  const callRoomId = useSelector(getAppCurrentCallRoomId)
  const { title } = useConfig(callRoomId)

  const onClickNotification = useCallback(() => {
    dispatch(switchRoomSubmit({ roomId: callRoomId }))
    navigate(SCREEN_ROOM_CALL, {
      roomId: callRoomId,
      shouldToggleMicrophoneAtInit: false,
    })
    dispatch(resetCallAutoEndTimer())
  }, [callRoomId, dispatch])

  const onClose = useCallback(() => {
    dispatch(resetCallAutoEndTimer())
  }, [dispatch])

  const showCallAutoEndBottomSheet = showAutoEndPopup && callOpened
  const showCallAutoEndNotification = showAutoEndPopup && !callOpened

  useEffect(() => {
    if (showCallAutoEndBottomSheet) {
      showBottomSheet({
        bottomSheetType: BottomSheetEnum.CallAutoEndPopup,
        bottomSheetOnDismissCallback: onClose,
      })
      return () => closeBottomSheet()
    }
  }, [onClose, showCallAutoEndBottomSheet])

  useEffect(() => {
    if (showCallAutoEndNotification) {
      showInfoNotifier(
        { title: title },
        {
          duration: 0,
          Component: CallAutoEndNotice,
          componentProps: {
            roomId: callRoomId,
          },
          onPress: onClickNotification,
          onHidden: onClose,
        },
      )
      return () => {
        Notifier.hideNotification()
      }
    }
  }, [
    callRoomId,
    onClickNotification,
    onClose,
    showCallAutoEndNotification,
    title,
  ])

  return null
})

interface CallAutoEndNoticeI {
  title: string
  roomId: string
}
const CallAutoEndNotice = ({ title, roomId }: CallAutoEndNoticeI) => {
  const styles = getStyles()
  const strings = useStrings()

  const minutes = useCountdown()

  return (
    <SafeAreaView>
      <View style={styles.container}>
        <RoomAvatarImage roomId={roomId} />
        <View style={styles.content}>
          <View style={s.centerAlignedRow}>
            <Text style={styles.title}>{title}</Text>
            <CloseButton />
          </View>
          <Text style={styles.description}>
            <SvgIcon
              name="phoneFilled"
              color={colors.white_snow}
              width={ICON_SIZE_16}
              height={ICON_SIZE_16}
              style={styles.icon}
            />
            {strings.call.callNoActivityNotice.replace('$0', String(minutes))}
          </Text>
        </View>
      </View>
    </SafeAreaView>
  )
}

const getStyles = createThemedStylesheet((theme) => {
  const styles = StyleSheet.create({
    container: {
      ...s.row,
      backgroundColor: theme.color.bg3,
      borderRadius: UI_SIZE_16,
      margin: UI_SIZE_12,
      padding: UI_SIZE_16,
    },
    content: {
      ...s.container,
      paddingLeft: UI_SIZE_8,
    },
    description: {
      ...theme.text.body,
      fontSize: 16,
    },
    icon: {
      marginRight: UI_SIZE_4,
    },
    title: {
      ...s.container,
      ...theme.text.bodySemiBold,
      fontSize: 16,
      marginBottom: UI_SIZE_4,
    },
  })
  return styles
})

export default CallAutoEndPopup
