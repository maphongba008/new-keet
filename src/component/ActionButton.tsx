import React from 'react'
import { StyleSheet, Text, View } from 'react-native'

import { ButtonBase } from 'component/Button'
import SvgIcon, { SvgIconType } from 'component/SvgIcon'
import { colors, createThemedStylesheet } from 'component/theme'
import s, { UI_SIZE_8, UI_SIZE_12, UI_SIZE_18 } from 'lib/commonStyles'

interface ActionButtonI {
  label: string
  iconLeft?: SvgIconType | React.ReactNode
  iconRight?: SvgIconType | React.ReactNode
  onPressItem: () => void
}

function ActionButton(props: ActionButtonI) {
  const { label, iconLeft, iconRight, onPressItem, ...rest } = props
  const styles = getStyles()

  return (
    <>
      <ButtonBase style={styles.actionBtn} onPress={onPressItem} {...rest}>
        <View style={[s.row, s.alignItemsEnd]}>
          {typeof iconLeft === 'string' ? (
            <SvgIcon
              name={iconLeft as SvgIconType}
              width={UI_SIZE_18}
              height={UI_SIZE_18}
              color={colors.white_snow}
            />
          ) : (
            iconLeft
          )}
          <View style={[s.row, s.alignItemsCenter, styles.textWrapper]}>
            <Text style={styles.btnTypeLabel}>{label}</Text>
          </View>
        </View>
        {typeof iconRight === 'string' ? (
          <SvgIcon
            name={iconRight as SvgIconType}
            width={UI_SIZE_18}
            height={UI_SIZE_18}
            color={colors.white_snow}
          />
        ) : (
          iconRight
        )}
      </ButtonBase>
    </>
  )
}

const getStyles = createThemedStylesheet((theme) => {
  const styles = StyleSheet.create({
    actionBtn: {
      ...s.row,
      ...s.alignItemsCenter,
      ...s.flexSpaceBetween,
      backgroundColor: theme.color.grey_600,
      borderRadius: UI_SIZE_12,
      height: 52,
      marginTop: theme.spacing.normal,
      padding: theme.spacing.standard,
    },
    btnTypeLabel: {
      ...theme.text.body,
      fontSize: 15,
    },
    textWrapper: {
      marginLeft: UI_SIZE_8,
    },
  })
  return styles
})

export default ActionButton
