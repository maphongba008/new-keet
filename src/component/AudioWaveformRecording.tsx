import React, { memo, useCallback, useEffect, useMemo, useRef } from 'react'
import { ScrollView, StyleSheet, View } from 'react-native'
import Reanimated, { ZoomIn } from 'react-native-reanimated'
import _isEmpty from 'lodash/isEmpty'

import { createThemedStylesheet } from 'component/theme'
import s, { UI_SIZE_4, UI_SIZE_40 } from 'lib/commonStyles'
import {
  WAVEFORM_HEIGHT,
  WAVEFORM_MIN_HEIGHT,
  WAVEFORM_SPACE,
} from 'lib/constants'
import { getMessageCellAvailableWidth } from 'lib/size'

const RECORDING_WAVEFORM_WIDTH = 2

type AudioWaveformRecordingProps = {
  samples: Array<number> | undefined
}

const AudioWaveformRecording = ({
  samples = [],
}: AudioWaveformRecordingProps) => {
  const styles = getStyles()
  const scrollRef = useRef<ScrollView>(null)

  const maxSampleCount = useMemo(() => {
    const availableWidth = getMessageCellAvailableWidth() - UI_SIZE_40
    return availableWidth / (RECORDING_WAVEFORM_WIDTH + WAVEFORM_SPACE * 2)
  }, [])

  useEffect(() => {
    if (samples.length % 2 !== 0 || samples.length < maxSampleCount) return

    scrollRef?.current?.scrollToEnd({ animated: true })
  }, [maxSampleCount, samples.length])

  const renderWaveform = useCallback(
    (sample: number, key: number) => (
      <Reanimated.View
        key={key}
        entering={ZoomIn}
        style={[
          styles.waveformItem,
          {
            height: Math.min(
              Math.max(sample / 10, WAVEFORM_MIN_HEIGHT),
              WAVEFORM_HEIGHT,
            ),
          },
        ]}
      />
    ),
    [styles.waveformItem],
  )

  if (_isEmpty(samples)) {
    return (
      <View style={[s.container, s.centeredLayout]}>
        <View style={styles.hLine} />
      </View>
    )
  }

  return (
    <ScrollView
      ref={scrollRef}
      style={styles.waveformSamplesContainer}
      contentContainerStyle={s.alignItemsCenter}
      showsHorizontalScrollIndicator={false}
      scrollEnabled={false}
      horizontal
    >
      <View style={[s.row, s.alignItemsCenter]}>
        {samples.map(renderWaveform)}
      </View>
    </ScrollView>
  )
}

export default memo(AudioWaveformRecording)

const getStyles = createThemedStylesheet((theme) => {
  const styles = StyleSheet.create({
    hLine: {
      backgroundColor: theme.background.bg_2,
      height: 1,
      position: 'absolute',
      width: '100%',
      zIndex: 0,
    },
    waveformItem: {
      backgroundColor: theme.color.blue_400,
      borderRadius: UI_SIZE_4,
      marginHorizontal: WAVEFORM_SPACE,
      width: RECORDING_WAVEFORM_WIDTH,
    },
    waveformSamplesContainer: {
      marginRight: UI_SIZE_4,
    },
  })
  return styles
})
