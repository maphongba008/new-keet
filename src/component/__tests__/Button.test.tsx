import React from 'react'
import { render, screen } from '@testing-library/react-native'

import { ButtonBase } from 'component/Button'

// refer to CloseButton tests for related button test cases
describe('ButtonBase component tests', () => {
  it('renders correctly with default props', () => {
    render(<ButtonBase />)
    expect(screen.toJSON()).toMatchSnapshot()
  })
  it('has correct testID prop when passed', () => {
    render(<ButtonBase testID="button_base" />)
    const button = screen.getByRole('button')
    expect(button).toHaveProp('testID', 'button_base')
  })
  it('renders styles with correct opacity if disabled', () => {
    render(<ButtonBase disabled testID="button_base" />)
    expect(screen.getByTestId('button_base')).toHaveStyle({
      opacity: 0.5,
    })
    expect(screen.toJSON()).toMatchSnapshot()
  })
})
