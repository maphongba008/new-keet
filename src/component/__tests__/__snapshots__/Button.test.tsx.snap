// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonBase component tests renders correctly with default props 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "opacity": 1,
      },
      undefined,
    ]
  }
/>
`;

exports[`ButtonBase component tests renders styles with correct opacity if disabled 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "opacity": 0.5,
      },
      undefined,
    ]
  }
  testID="button_base"
/>
`;
